bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/HeapSizeManager.java

```diff
@@ -64,60 +74,79 @@ public long getMaxHeapSize() {
   }
 
   public synchronized void waitUntilAllOperationsAreDone() throws InterruptedException {
+    isFlushing.set(true);
     boolean performedWarning = false;
-    while(!pendingOperationsWithSize.isEmpty()) {
+    while (true) {
       cleanupFinishedOperations();
       if (pendingOperationsWithSize.isEmpty()) {
         break;
       }
-      if (!performedWarning
-          && lastOperationChange + INTERVAL_NO_SUCCESS_WARNING < System.currentTimeMillis()) {
-        long lastUpdated = (System.currentTimeMillis() - lastOperationChange) / 1000;
-        LOG.warn("No operations completed within the last %d seconds."
-            + "There are still %d operations in progress.", lastUpdated,
-          pendingOperationsWithSize.size());
+      if (!performedWarning && requiresWarning()) {
+        warn();
         performedWarning = true;
-        waitForCompletions(FINISH_WAIT_MILLIS);
       }
+      wait(FINISH_WAIT_MILLIS);
     }
     if (performedWarning) {
       LOG.info("flush() completed");
     }
+    isFlushing.set(false);
+  }
+
+  private boolean requiresWarning() {
+    return lastOperationChange + INTERVAL_NO_SUCCESS_WARNING < System.currentTimeMillis();
   }
 
+  private void warn() {
+    long lastUpdated = (System.currentTimeMillis() - lastOperationChange) / 1000;
+    LOG.warn(
+        "No operations completed within the last %d seconds."
+        + "There are still %d operations in progress.",
+        lastUpdated, pendingOperationsWithSize.size());
+  }
 
   private void cleanupFinishedOperations() {
```

Comment:
Can we do this?

List<Long> toClean = new ArrayList<>();
completedOperationIds.drainTo(toClean);
for (Long operationSequenceId : toClean) {
  markOperationComplete(operationSequenceId);
}
if (!toClean.empty()) {
  lastOperationChange = System.currentTimeMillis();
}